fix: defaultHomePagePath to be last visited page or alphatically first active object with the name

Nabhag8848 opened the pull request:
### ISSUE 

- Closes #6612
- Closes #6125
- Closes #5949
- Closes #6652 

### Description 

- [x] need to check changes in jest test.
- [x] fallback to alphabetically firstActiveObject with the name if no last visited exist

https://github.com/user-attachments/assets/dd11480b-c47f-4393-9857-8a55467061e3

-  [x] fallback to last visited page with the last visited view by default if no views would have toggled with subNav or viewChangeDropdown it will fallback to INDEX or if no INDEX view then zero position view, works with both subNavViewBar and viewChangeDropdown. 


https://github.com/user-attachments/assets/33e97e55-2aa2-4c45-a3ab-fc8e43f4964c


https://github.com/user-attachments/assets/d1db76a2-da59-4cd2-81bf-d6119408fbbf

- [x] lastVisited view across the objects have been persisted so now navigating back from x object to y or z to x will open always last visited view and defaults to index or zero position view. 


https://github.com/user-attachments/assets/70a01a11-a7ef-4031-926e-02923551466c

- [x] lastVisited Page with view has been persisted across the workspace, scope is per workspace so jumping between workspace will also work to have lastvisited object of particular workspace. 


https://github.com/user-attachments/assets/25107339-8ec1-4421-9f6e-1da43b8f4816

- [x] when lastVisitedObject has been deactivated and going back from settings should have a fallback Object that is alphabetically First activeObject. 


https://github.com/user-attachments/assets/6b24a933-b139-49ac-82b2-eac5e4848516


- [x] Creation of new View of **anyType** should also get persist and that should get lastVisitedObject with View in case the user leaves from there right away.


https://github.com/user-attachments/assets/80ff7114-051d-4e9b-ab58-0e1e3a7d328c

- [x] Similarly deleted view also works. 

https://github.com/user-attachments/assets/cb0b8043-fba4-4a66-941d-b3fa0a57eb22


- [x] fixed active subnav background when opening object directly with root path **/** , it wasn't showing active subNav background. 

Before: 

https://github.com/user-attachments/assets/db341c4a-f1f9-43c4-9838-37d1a1f5ab8e

Fixed: 

https://github.com/user-attachments/assets/0f0fd492-bc5d-4efe-b695-bee4e3f41d4e

Nabhag8848 commented on packages/twenty-front/src/modules/navigation/hooks/useLastVisitedPageOrView.ts:
```diff
@@ -0,0 +1,92 @@
+      objectMetadataId,
+    );
+
+    const [newFallbackObjectMetadataItem] =
+      alphaSortedActiveObjectMetadataItems.filter(
+        (item) => item.id !== objectMetadataId,
+      );
```
we are assuming there would be atleast one active object as of now.

lucasbordeau commented:
When I delete the local storage to simulate a first time load, it goes on the companies page (which is the first alphabetically sorted object), but there's a mismatch between the selected view in the left menu and in the view button : 

<img width="413" alt="Capture d’écran 2024-08-21 à 15 16 06" src="https://github.com/user-attachments/assets/b177627b-6096-4bf2-83f1-cd3e8d4e3c4e">

lucasbordeau commented:
Could you add a small gap between the two sub nav items ? (Try 2 o 4px)

lucasbordeau commented on packages/twenty-front/src/modules/object-metadata/utils/getObjectMetadataAndViewIdFromPath.ts:
```diff
@@ -0,0 +1,15 @@
+  }
+
+  return {
+    componentId: result[1],
```
componentId => objectNamePlural ?

Nabhag8848 replied:
Won't require anymore, same reason as we changed to have in QueryParamsViewIdEffect.

lucasbordeau commented on packages/twenty-front/src/modules/object-metadata/utils/getObjectMetadataAndViewIdFromPath.ts:
```diff
@@ -0,0 +1,15 @@
+const EXTRACT_OBJECT_VIEWID_PLURAL_NAME =
+  /\/objects\/([^/?]+)(?:\?view=([^&]+))?/;
+
+export const getObjectAndViewIdFromPath = (path: string) => {
```
getObjectNamePluralAndViewIdFromPath ?

Nabhag8848 replied:
Won't require anymore, same reason as we changed to have in QueryParamsViewIdEffect.

lucasbordeau commented on packages/twenty-front/src/modules/settings/data-model/object-details/components/SettingsObjectSummaryCard.tsx:
```diff
@@ -38,15 +39,18 @@ export const SettingsObjectSummaryCard = ({
   const theme = useTheme();
   const { getIcon } = useIcons();
   const Icon = getIcon(iconKey);
+  const objectMetadataId = objectMetadataItem.id;
```
objectMetadataItemId

lucasbordeau commented on packages/twenty-front/src/modules/ui/navigation/navigation-drawer/components/NavigationDrawerItem.tsx:
```diff
@@ -147,7 +149,17 @@ export const NavigationDrawerItem = ({
-    if (isNonEmptyString(to)) navigate(to);
+    if (isNonEmptyString(to)) {
+      const objectAndViewId = getObjectAndViewIdFromPath(to);
+      if (!isNull(objectAndViewId)) {
```
!isDefined

Nabhag8848 replied:
Won't require anymore, same reason as we changed to have in QueryParamsViewIdEffect.

lucasbordeau commented on packages/twenty-front/src/modules/ui/navigation/navigation-drawer/components/NavigationDrawerItem.tsx:
```diff
@@ -147,7 +149,17 @@ export const NavigationDrawerItem = ({
+      const objectAndViewId = getObjectAndViewIdFromPath(to);
+      if (!isNull(objectAndViewId)) {
+        const { componentId, viewId } = objectAndViewId;
+        setLastVisitedObjectOrView({
```
Here we would set both new atoms instead of this single atom

lucasbordeau commented on packages/twenty-front/src/modules/views/view-picker/hooks/useViewPickerPersistView.ts:
```diff
@@ -20,9 +21,10 @@ export const useViewPickerPersistView = () => {
   const { createView, selectView, removeView, updateView } = useHandleViews();
 
-  const { viewsOnCurrentObject } = useGetCurrentView();
+  const { viewsOnCurrentObject, componentId } = useGetCurrentView();
```
componentId shouldn't be used like that if what we intend is to use objectNamePlural, doesn't matter if componentId === objectNamePlural

lucasbordeau replied:
You can even modify useGetCurrentView to return objectNamePlural

lucasbordeau commented on packages/twenty-front/src/modules/views/view-picker/hooks/useViewPickerPersistView.ts:
```diff
@@ -77,11 +85,22 @@ export const useViewPickerPersistView = () => {
           )[0].id,
         );
         await removeView(viewPickerReferenceViewId);
+        const [indexOrFirstView] = viewsOnCurrentObject
```
Could this be turned into a util ? getFirstAvailableView(referenceViewId?: string)

Nabhag8848 replied:
Won't require anymore, as we now have single source of truth. now we have setters directly in QueryParamsViewIdEffect, will write description once i make all changes. Thanks.

Nabhag8848 commented on packages/twenty-front/src/modules/navigation/hooks/useLastVisitedPageOrView.ts:
```diff
@@ -0,0 +1,97 @@
+    viewId: string;
+  }) => {
+    const fallbackObjectMetadataId =
+      findActiveObjectMetadataItemBySlug(componentId)?.id ?? '';
```
that would never happen unless, user visits object directly through link which is deactivated **when visited** and In this case I've made changes that only if fallbackObjectMetadataId exists and its not an empty string then only set and have localStorageEffect.

Nabhag8848 commented:
made all changes accordingly, have few more questions, writing...
- [x] Split Atoms
- [x] Fixed Naming Convention 
- [x] Split Hooks as per clean code guidelines
- [x] subNavItems gap of 2px lgtm.
- [x] MAJOR CHANGE: We now have single source of truth, previous solution wasn't scalable, and having this changes and setting atom values within UI layer didn't make sense, now we set lastVisitedViewState and  lastVisitedObjectMetadataItemState right in  `QueryParamsViewIdEffect`. this solves all our edge cases and in near future we would require very less changes for this. 
- [x] It fixes SignIn Issue #6652 we were getting mockObjects on first signin completion, which is solved now.

lucasbordeau commented on packages/twenty-front/src/hooks/useDefaultHomePagePath.tsx:
```diff
@@ -35,24 +36,26 @@ export const useDefaultHomePagePath = () => {
   const getFirstObjectInfo = () => {
     const [metadata] = alphaSortedActiveObjectMetadataItems;
```
metadata => firstObjectMetadataItem

lucasbordeau reviewed (commented):
Way more clear thanks

Nabhag8848 commented:
@lucasbordeau my last changes before main merge included a minor mistake, the view from  getViewMatchingObjectMetdataItemId should have passed firstObjectMetadataItem.id instead.

lucasbordeau commented:
@Nabhag8848 Could you please rework useDefaultHomePagePath to avoid having a let variable ? There are too many utils, you can replace them with useMemo and compute everything and finish by using a ternary to know which one you want to return.

Nabhag8848 commented:
@lucasbordeau made changes, thanks !

lucasbordeau commented:
@Nabhag8848 Nice work, I added some minor fix. 

There's still one test not passing. Otherwise it's ready to merge !

Nabhag8848 commented:
Hey @lucasbordeau test: `should return proper path when currentUser is defined and view exists` has been fixed. 
there is other two around exhaustive test but I see its in current main aswell.

Nabhag8848 commented:
also few questions not related to PR: 

- Can you share good resources/repository, to learn a good practices around test driven development in react ? 
- When do we generally make realease ? 
- Also is there any integration work that users are asking and which we should prioritise ? - I just show in discussion there were many asking for n8n, is anything blocking for same at the moment ? (btw - I love building integrations and have build bunch in past).

lucasbordeau commented:
For TDD in React I recommend two rules of thumb : 
- Doing it for critical utils like date / string / number manipulation, so this would be with unit test
- Doing it for critical or extensively used UI components with Storybook stories

lucasbordeau commented:
For your other questions I recommend that you contact @charlesBochet on Discord ;)

Nabhag8848 commented:
Thanks !

Bonapara commented:
@Nabhag8848 just opened this one which is related -> https://github.com/twentyhq/twenty/issues/6786
